synapse/tests/test_cortex.py

```diff
@@ -1331,6 +1291,61 @@ def test_cortex_int_indx(self):
             self.len(0, core.eval('testint<20'))
             self.len(1, core.eval('testint<30'))
 
+    def test_cortex_ontag(self):
+
+        with self.getTestCore() as core:
+
+            tags = {}
+            def onadd(node, tag, valu):
+                tags[tag] = valu
+
+            def ondel(node, tag, valu):
+                self.none(node.getTag(tag))
+                self.false(node.hasTag(tag))
+                tags.pop(tag)
+
+            core.onTagAdd('foo', onadd)
+            core.onTagAdd('foo.bar', onadd)
+            core.onTagAdd('foo.bar.baz', onadd)
+
+            core.onTagDel('foo', ondel)
+            core.onTagDel('foo.bar', ondel)
+            core.onTagDel('foo.bar.baz', ondel)
+
+            with core.snap() as snap:
+
+                node = snap.addNode('teststr', 'hehe')
+                node.addTag('foo.bar.baz', valu=(200, 300))
+
+                self.eq(tags.get('foo'), (None, None))
+                self.eq(tags.get('foo.bar'), (None, None))
+                self.eq(tags.get('foo.bar.baz'), (200, 300))
+
+                node.delTag('foo.bar')
+
+                self.eq(tags.get('foo'), (None, None))
+
+                self.none(tags.get('foo.bar'))
+                self.none(tags.get('foo.bar.baz'))
+
+    def test_cortex_del_univ(self):
+
+        with self.getTestCore() as core:
+
+            core.model.addUnivProp('hehe', ('int', {}), {})
+
+            self.len(1, core.eval('[ teststr=woot .hehe=20 ]'))
+            self.len(1, core.eval('.hehe'))
+            self.len(1, core.eval('.hehe [ -.hehe ]'))
+            self.len(0, core.eval('.hehe'))
+
+    def test_cortex_snap_eval(self):
```

Comment:
this is a merge error - most of the content in `test_cortex_snap_eval` came from `test_cortex_int_indx`